### PR TITLE
feat: extract node version

### PR DESCRIPTION
**Pull request by karniwl**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
Extracts node version of docker image by running `node --version`.

### CODE REVIEW

File: lib/analyzer/engines-analyzer.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    Image: targetImage,
+    AnalyzeType: 'engines',
+    Analysis: engines,
+  };
+}
+
+async function getEngines(targetImage: string) {
+  const engines: Engine[] = [];
+  const node = await getNodeEngine(targetImage);
+  if (node) { engines.push(node); }
```
Comment (orkamara) [inline lib/analyzer/engines-analyzer.ts]:
```suggestion
  if (node) {
  engines.push(node);
  }
```

File: lib/analyzer/engines-analyzer.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+async function analyze(targetImage: string) {
+  const engines = await getEngines(targetImage);
+  return {
+    Image: targetImage,
+    AnalyzeType: 'engines',
+    Analysis: engines,
+  };
+}
+
+async function getEngines(targetImage: string) {
```
Comment (orkamara) [inline lib/analyzer/engines-analyzer.ts]:
```suggestion
async function getEngines(targetImage: string): Promise<Engine[]> {
```

Comment (orkamara) [reply lib/analyzer/engines-analyzer.ts]: Help us for catching types errors:)

File: lib/analyzer/engines-analyzer.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  const engines: Engine[] = [];
+  const node = await getNodeEngine(targetImage);
+  if (node) { engines.push(node); }
+  return engines;
+}
+
+function getNodeEngine(targetImage: string) {
```
Comment (orkamara) [inline lib/analyzer/engines-analyzer.ts]:
```suggestion
function getNodeEngine(targetImage: string): Promise<Engine | null> {
```

File: lib/analyzer/engines-analyzer.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      if (typeof stderr === 'string' && stderr.indexOf('not found') >= 0) {
+        return '';
+      }
+      throw new Error(stderr);
+    })
+    .then(parseNodeEngine);
+}
+
+function parseNodeEngine(version: string) {
+  const eng: Engine = {
```
Comment (orkamara) [inline lib/analyzer/engines-analyzer.ts]: did you check different types of node versions \ images?

File: lib/analyzer/engines-analyzer.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    name: 'node',
```
Comment (orkamara) [inline lib/analyzer/engines-analyzer.ts]: you should add a validation for the output - it's not enough to assume that it's a valid one)
consider using a regexp for that? (x.y.z - you can remove the leading v)

File: lib/analyzer/engines-analyzer.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    version: version.trim(),
```
Comment (orkamara) [inline lib/analyzer/engines-analyzer.ts]: please remove the leading v (`v8.11.3`)

File: lib/analyzer/engines-analyzer.ts (hunk 4 above)
Comment (karniwl) [reply lib/analyzer/engines-analyzer.ts]: yes

File: lib/analyzer/engines-analyzer.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import { Engine } from './types';
+
+const semver = require('semver');
+
+export {
+  analyze,
+};
+
+async function analyze(targetImage: string) {
+  const engines = await getEngines(targetImage);
```
Comment (orkamara) [inline lib/analyzer/engines-analyzer.ts]: Sorry for that, but please rename all `engines` to `binaries` (sorry)

File: lib/analyzer/engines-analyzer.ts [hunk 8]
```diff
@@ -33,9 +37,11 @@ function getNodeEngine(targetImage: string) {
 }
 
 function parseNodeEngine(version: string) {
+  const nodeVersion = semver.valid(version.trim());
```
Comment (orkamara) [inline lib/analyzer/engines-analyzer.ts]: Maybe:

```const nodeVersion = semver.valid(version.trim());

  if (!nodeVersion) {
    return null;
  }

  return {
    name: 'node',
    version: nodeVersion,
  };```

File: test/lib/analyzer/engines-analyzer.test.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    {
+      description: 'no Node in image',
+      targetImage: 'alpine:2.6',
+      enginesOutputLines: [''],
+      expectedEngines: { },
+    },
+    {
+      description: 'Node is in image',
+      targetImage: 'node:6.15.1',
+      enginesOutputLines: ['6.15.1'],
```
Comment (orkamara) [inline test/lib/analyzer/engines-analyzer.test.ts]:
```suggestion
      enginesOutputLines: ['v6.15.1'],
```

File: test/lib/analyzer/engines-analyzer.test.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+import { test } from 'tap';
+import * as sinon from 'sinon';
+
+import * as subProcess from '../../../lib/sub-process';
+import * as analyzer from '../../../lib/analyzer/engines-analyzer';
+
+test('analyze', async t => {
+
+  const examples = [
```
Comment (orkamara) [inline test/lib/analyzer/engines-analyzer.test.ts]: add anothre example for junk `enginesOutputLines` (vvadsf23123)

**Pull request merged**